feat(press): merge pressurization and air conditioning system

mjuhe opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
This PR merges the pressurization and air conditioning systems into one (ATA21). It creates a single cabin simulation, so pressure changes affect temperature and temperature changes affect pressure. This PR is a big refactor that intends to improve the press system and make it scalable for future implementations (aka for the A380).

Most of the changes will be completely unnoticable to the end user, but here is a list of the main ones:
- [x] Pressurization system and pressurization overhead moved to the aircraft specific portion of the code
- [x] RCPU system moved to aircraft specific portion of the code
- [x] Adjusted the volume of the pressurized cabin from references
- [x] One single air simulation for the cabin: so changing cabin temperature affects cabin pressure, and changing pressure affects temperature
- [x] Moved all aircraft specific constants out of `systems`, passed as generic argument
- [x] Passenger quantity and number of open doors are now read in the aircraft specific portion of code
- [x] Opening doors affects cabin pressure
- [x] Improved outflow valve flow out calculation (separate functions for supersonic and subsonic flow)
- [x] Added baro correction value to ADIRS
- [x] Connected ADIRS as input to CPCs (Arinc429)
- [x] Changed LVAR used for FMS landing elev to `FM1_LANDING_ELEVATION`
- [x] Fixed landing and departure cabin altitude calculation
        - When in auto, it uses FMS landing elev -> baro correction -> standard
        - When in manual, it uses standard only
- [x] Added outflow valve motor and improved overall OFV code design
- [x] Connected outflow valve to electrical system
- [x] The cabin vertical speed transmitted to the ECAM is limitted to +/- 6400fpm as in the real aircraft
- [x] Many many other refactor improvements 😄 


## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

## References
<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->
The usual: FCOM, AMM, and others.

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->
This is a big refactor, but there are still things to do!
Future work:
- Connect CPCs to electrical system.
- Failures.
- Use ARINC for outputs.
- Use a single LVAR for landing elevation everywhere (now two are being used)
- Connect the baro correction in the ADIRS to the PDF

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): Commander Ga#1548

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
As mentioned, not a lot of noticeable changes in this PR so the objective is to ensure the pressurization system still works normally. You can use the testing instructions in #5159 and #5504 as guideline, but you don't have to follow every single point there.

The most meaningful change introduced with this PR is the calculation of cabin altitude based on local QNH when in auto mode. Validate the cabin altitude is calculated as expected (see table below). Note! If you have packs on and all doors closed the cabin altitude might look slightly lower than expected (50-100ft), this is normal.

![image](https://user-images.githubusercontent.com/60938089/219775113-95b238c3-0e9d-42c8-9166-9b85aac9748f.png)


<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

2hwk commented on fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning.rs:
```diff
@@ -40,131 +55,2155 @@ impl A380AirConditioning {
-    // TODO: Improve volume according to specs
-    const A380_CABIN_VOLUME_CUBIC_METER: f64 = 200.; // m3
-    const A380_COCKPIT_VOLUME_CUBIC_METER: f64 = 10.; // m3
+impl A320Cabin {
```
```suggestion
impl A320Cabin {
```
Shouldn't naming still be A380

mjuhe replied:
Yeah I left it as A320 to signify that it still hasn't been adapted to the A380. It's just a naming thing though so I can change it if you think it's better.

mjuhe commented on fbw-common/src/wasm/systems/systems/src/shared/mod.rs:
```diff
@@ -206,8 +222,11 @@ pub trait PackFlowValveState {
 
-pub trait GroundSpeed {
-    fn ground_speed(&self) -> Velocity;
+pub trait AdirsSignalInterface {
```
I put it in `shared` in case it was useful for other systems in the future 😃, but no I agree at the moment it's only used in the pressurization system. I've renamed it and moved it to `air_conditioning` as suggested.

mjuhe commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/pressure_valve.rs:
```diff
@@ -0,0 +1,462 @@
+        Self {
+            open_amount: Ratio::new::<percent>(0.),
+        }
+    }
```
Absolutely, this is from before the refactor because they stored the travel time as well. Changed now to `new_open()` and `new_closed()`.

mjuhe commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -0,0 +1,1754 @@
+    manual_to_auto_switch: bool,
+    cabin_temperature: ThermodynamicTemperature,
+
+    _constants: Constants<C>,
```
Super cool, thanks! I've learned something new 😄.

tracernz commented on fbw-common/src/wasm/systems/systems/src/navigation/adirs.rs:
```diff
@@ -432,9 +443,18 @@ impl SimulationElement for AirDataInertialReferenceSystem {
-    fn ground_speed(&self) -> Velocity {
-        self.adirus[0].ground_speed()
+impl AdirsToAirCondInterface for AirDataInertialReferenceSystem {
+    fn ground_speed(&self, adiru_number: usize) -> Velocity {
```
You need an Arinc429Word here?

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -0,0 +1,1741 @@
+    impl TestAdirs {
+        fn new() -> Self {
+            Self {
+                true_airspeed: Velocity::new::<knot>(0.),
```
```suggestion
                true_airspeed: Velocity::default(),
```

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -0,0 +1,1741 @@
+    }
+    impl AdirsToAirCondInterface for TestAdirs {
+        fn ground_speed(&self, _adiru_number: usize) -> Arinc429Word<Velocity> {
+            Arinc429Word::new(Velocity::new::<knot>(0.), SignStatus::NormalOperation)
```
```suggestion
            Arinc429Word::new(Velocity::default(), SignStatus::NormalOperation)
```

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -0,0 +1,1741 @@
+            Arinc429Word::new(self.true_airspeed, SignStatus::NormalOperation)
+        }
+        fn baro_correction(&self, _adiru_number: usize) -> Arinc429Word<Pressure> {
+            Arinc429Word::new(Pressure::new::<hectopascal>(0.), SignStatus::NoComputedData)
```
```suggestion
            Arinc429Word::new(Pressure::default(), SignStatus::NoComputedData)
```

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -0,0 +1,1741 @@
+        fn new() -> Self {
+            Self {
+                duct_demand_temperature: ThermodynamicTemperature::new::<degree_celsius>(24.),
+                pack_flow: MassRate::new::<kilogram_per_second>(0.),
```
```suggestion
                pack_flow: MassRate::default(),
```

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -0,0 +1,1741 @@
+                ),
+                safety_valve: SafetyValve::new(),
+                press_overhead: TestPressurizationOverheadPanel::new(context),
+                engine_1: TestEngine::new(Ratio::new::<percent>(0.)),
+                engine_2: TestEngine::new(Ratio::new::<percent>(0.)),
```
```suggestion
                engine_1: TestEngine::new(Ratio::default()),
                engine_2: TestEngine::new(Ratio::default()),
```

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/acs_controller.rs:
```diff
@@ -1821,11 +1976,13 @@ mod acs_controller_tests {
         }
 
         fn update_after_power_distribution(&mut self, context: &UpdateContext) {
+            let lgciu_gears_compressed = self.lgciu1.compressed() & self.lgciu2.compressed();
```
```suggestion
            let lgciu_gears_compressed = self.lgciu1.compressed() && self.lgciu2.compressed();
```

Gurgel100 reviewed (changes requested):
Great work. This is quite a PR. I didn't comment on air_conditioning.rs of the A380 because it's essentially the same the A320 one.

mjuhe commented on fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning.rs:
```diff
@@ -2,32 +2,47 @@ use systems::{
+        InitContext, Read, SimulationElement, SimulationElementVisitor, SimulatorReader,
+        SimulatorWriter, UpdateContext, VariableIdentifier, Write,
     },
-    simulation::{InitContext, SimulationElement, SimulationElementVisitor, UpdateContext},
 };
-use uom::si::{f64::*, mass_rate::kilogram_per_second, volume::cubic_meter};
+
+use std::time::Duration;
+use uom::si::{f64::*, pressure::hectopascal, ratio::percent, velocity::knot};
 
 pub(super) struct A380AirConditioning {
-    a380_cabin: A380Cabin,
-    a380_air_conditioning_system: AirConditioningSystem<3, 2, 4>,
+    a320_cabin: A320Cabin,
+    a320_air_conditioning_system: AirConditioningSystem<3, 2, 4>,
+    a320_pressurization_system: A320PressurizationSystem,
```
Yes, all of these will be replaced with their corresponding A380 systems when modelled.

mjuhe commented on fbw-a32nx/src/wasm/systems/a320_systems/src/air_conditioning.rs:
```diff
@@ -93,84 +123,2145 @@ impl SimulationElement for A320AirConditioning {
+            .iter_mut()
+            .any(|controller| controller.should_switch_cpc())
+        {
+            self.active_system = if self.active_system == 1 { 2 } else { 1 };
```
Thanks for the suggestion but I think it's a bit more readable the way it is now :).

mjuhe commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/acs_controller.rs:
```diff
@@ -144,6 +147,20 @@ impl<const ZONES: usize, const ENGINES: usize> AirConditioningSystemController<Z
 
+    fn ground_speed(&self, adirs: &impl AdirsToAirCondInterface) -> Option<Velocity> {
+        // TODO: Verify ADIRU check order
+        let mut adiru_check_order = [1, 2, 3].iter();
+        loop {
+            let adiru_number = adiru_check_order.next().unwrap_or(&99_usize);
+            if adiru_number == &99_usize {
+                return None;
+            }
+            if let Some(data) = adirs.ground_speed(*adiru_number).normal_value() {
+                return Some(data);
+            }
+        }
```
Thanks for the suggestion, this is indeed so much better!

mjuhe commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_air.rs:
```diff
@@ -1,70 +1,353 @@
+    for CabinAirSimulation<C, ZONES>
+{
+    fn cabin_temperature(&self) -> Vec<ThermodynamicTemperature> {
+        let mut cabin_temperature_vector = Vec::new();
+        for zone in self.cabin_zones.iter() {
+            cabin_temperature_vector.append(&mut zone.cabin_temperature())
+        }
+        cabin_temperature_vector
```
I love this one too, thanks! I've used `flat_map()` instead of `map()` and then `flatten()` though as the docs say it's preferable.

mjuhe commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -0,0 +1,1741 @@
+
+            landing_elevation: Length::new::<foot>(0.),
+            departure_elevation: Length::new::<foot>(0.),
+            destination_qnh: Pressure::new::<hectopascal>(0.),
```
I left the safety valve initiation as `Ratio::new::<percent>(0.)` for clarity. Changed all others as suggested.

mjuhe commented on fbw-a32nx/src/wasm/systems/a320_systems/src/air_conditioning.rs:
```diff
@@ -93,84 +123,2145 @@ impl SimulationElement for A320AirConditioning {
+                .then()
+                .command_aircraft_climb(Length::new::<foot>(0.), Length::new::<foot>(10000.))
+                .with()
+                .ambient_pressure_of(Pressure::new::<hectopascal>(696.85))
```
Removed as `command_aircraft_climb` already modifies ambient pressure.

mjuhe commented on fbw-a32nx/src/wasm/systems/a320_systems/src/air_conditioning.rs:
```diff
@@ -93,84 +123,2145 @@ impl SimulationElement for A320AirConditioning {
+                    .on_ground()
+                    .run_and()
+                    .command_packs_on_off(false)
+                    .ambient_pressure_of(Pressure::new::<hectopascal>(1020.)) // Equivalent to 10,000ft from tables
```
No, you are right, comment should not have been there!

mjuhe commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -0,0 +1,1754 @@
+        safety_valve: &SafetyValve,
+    ) {
+        self.exterior_pressure = cabin_simulation.exterior_pressure();
+        self.exterior_vertical_speed = context.vertical_speed();
```
Regarding exterior pressure = yes, you are correct, this is read from the ADIRS, not measured directly by the CPCs, I've fixed it now. Good spot thanks.

Regarding exterior vertical speed = this is not read from the ADIRS in the real thing. The CPCs calculate the vertical speed from the readings of ambient pressure. I was using `context` for convenience, but I've now modified it so it's calculated from the pressure readings to represent better how it happens in the real system. Now `context` is only used to read `delta`, everything else is taken from the correct systems.

mjuhe commented on fbw-a32nx/src/wasm/systems/a320_systems/src/air_conditioning.rs:
```diff
@@ -93,84 +123,2145 @@ impl SimulationElement for A320AirConditioning {
-                ),
-            ],
+            passenger_rows_id,
+            fwd_door_id: context.get_identifier(Self::FWD_DOOR.to_owned()),
+            rear_door_id: context.get_identifier(Self::REAR_DOOR.to_owned()),
```
It does work, this has been in master for a while 😃. It's the same way the ADIRS read aircraft variables. I would rather not include it in `UpdateContext` as I don't think it's something that will be needed across multiple systems.

mjuhe commented:
Thank you @2hwk @BlueberryKing @tracernz and @Gurgel100 for the extensive and excellent reviews. I keep learning more and more with every PR and I really do appreciate all the help and suggestions!

I've addressed all findings now.

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/acs_controller.rs:
```diff
@@ -3325,18 +3293,15 @@ mod acs_controller_tests {
-                test_bed.pack_flow(),
-                MassRate::new::<kilogram_per_second>(0.),
-            );
+            assert_eq!(test_bed.pack_flow(), MassRate::default(),);
```
```suggestion
            assert_eq!(test_bed.pack_flow(), MassRate::default());
```

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -127,11 +134,9 @@ impl<C: PressurizationConstants> CabinPressureController<C> {
         safety_valve: &SafetyValve,
     ) {
-        let (adirs_airspeed, adirs_ambient_pressure): (Option<Velocity>, Option<Pressure>) =
+        let (adirs_airspeed, _): (Option<Velocity>, Option<Pressure>) =
```
```suggestion
        let (adirs_airspeed, _) =
```

Gurgel100 commented on fbw-common/src/wasm/systems/systems/src/air_conditioning/cabin_pressure_controller.rs:
```diff
@@ -189,34 +193,48 @@ impl<C: PressurizationConstants> CabinPressureController<C> {
+        context: &UpdateContext,
+        adirs: &impl AdirsToAirCondInterface,
+    ) {
+        let (_, adirs_ambient_pressure): (Option<Velocity>, Option<Pressure>) =
```
```suggestion
        let (_, adirs_ambient_pressure) =
```

mjuhe commented on fbw-a32nx/src/wasm/systems/a320_systems/src/air_conditioning.rs:
```diff
@@ -1981,8 +2039,10 @@ mod tests {
                 .then()
-                .ambient_pressure_of(Pressure::new::<hectopascal>(465.63))
-                .vertical_speed_of(Velocity::new::<foot_per_minute>(0.))
+                .ambient_pressure_of(InternationalStandardAtmosphere::pressure_at_altitude(
+                    Length::new::<foot>(20000.),
+                ))
```
No, good spot. Removed now.

Gurgel100 commented on fbw-a32nx/src/wasm/systems/a320_systems/src/air_conditioning.rs:
```diff
@@ -1086,7 +1086,7 @@ mod tests {
 
         fn set_pressure_based_on_vs(&mut self, alt_diff: Length) {
-            // We find the atmosferic pressure that would give us the desired v/s
+            // We find the atmospheric  pressure that would give us the desired v/s
```
```suggestion
            // We find the atmospheric pressure that would give us the desired v/s
```

Gurgel100 commented on fbw-a380x/src/wasm/systems/a380_systems/src/air_conditioning.rs:
```diff
@@ -1092,7 +1092,7 @@ mod tests {
 
         fn set_pressure_based_on_vs(&mut self, alt_diff: Length) {
-            // We find the atmosferic pressure that would give us the desired v/s
+            // We find the atmospheric  pressure that would give us the desired v/s
```
```suggestion
            // We find the atmospheric pressure that would give us the desired v/s
```

Gurgel100 reviewed (commented):
Sorry one last thing ;)

Gurgel100 reviewed (approved):
Great work

tracernz reviewed (approved):
Brilliant!

tracernz commented on fbw-common/src/wasm/systems/systems/src/navigation/adirs.rs:
```diff
@@ -1962,6 +2059,38 @@ mod tests {
             ))
         }
 
+        fn baro_reference_1_hpa(&mut self, adiru_number: usize) -> Arinc429Word<f64> {
```
Why "reference" on these ones. Better to keep the nomenclature consistent throughout I think.

mjuhe replied:
Agreed :) Fixed now.

tracernz commented:
Quality Assurance Report

Discord: tracernz#3313
Object of testing: https://github.com/flybywiresim/a32nx/pull/7804
Tier of Testing: 1
Date: 5/3/2023

Testing Process:
Full flight from NZWN to NZCH.

Noted the cab V/S up/down arrows on the SD cruise page being quite active during cruise (FL220) while rate remained 0. Outflow valve was amber during approach. Otherwise performed okay.

Negatives:
- Double check threshold for up/down arrow.
- Outflow valve amber in-flight with no failures.

![image](https://user-images.githubusercontent.com/9995998/222951177-fe6713e9-995c-42d4-a685-a372f4929c1d.png)

Testing Results: Not Passed

mjuhe commented:
Thanks for the test @tracernz, I saw the same issue with the outflow valve opening before touchdown in my own testing but I thought I'd fixed it. Looks like I didn't. I won't have access to my simulator until next weekend so this will have to wait a few days but I'll have another look.

mjuhe commented:
@tracernz 
> Negatives:
> 
>     * Double check threshold for up/down arrow.
> 
>     * Outflow valve amber in-flight with no failures.
> 

Thanks again for the test. Both issues should be fixed now:

* Cruise page arrow: the problem was with the CPC state determination, it was never getting into cruise phase because the aircraft VS calculation was never less than |100fpm| when flying low above mountains. I've added a low pass filter now that fixes the issue.
* Ofv opening in descent: the landing elevation from the FMC is an ARINC word that I was taking as plain value by mistake. This broke the descent profile and the OFV opened when delta P was 0. Also fixed now.

mico975 commented:
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #7804
Tier of Testing  : 2
Date             : 15/03/2023

Testing Process: 
Flight LFMP - LEGE

As mentioned, not a lot of noticeable changes in this PR so the objective is to ensure the pressurization system still works normally. You can use the testing instructions in https://github.com/flybywiresim/a32nx/pull/5159 and https://github.com/flybywiresim/a32nx/pull/5504 as guideline, but you don't have to follow every single point there. - **OK**

The most meaningful change introduced with this PR is the calculation of cabin altitude based on local QNH when in auto mode. Validate the cabin altitude is calculated as expected (see table below). Note! If you have packs on and all doors closed the cabin altitude might look slightly lower than expected (50-100ft), this is normal. - **OK**

Negatives:       
N/A

Testing Results: 
Passed, issues fixed

Conclusions: 
LGTM

Media:       
![image](https://user-images.githubusercontent.com/78908672/225301069-5b44bc74-1997-41e5-99e1-adbe54c4b2f5.png)
